tests/test_trainer.py

```diff
@@ -617,3 +629,48 @@ def assert_flos_extraction(trainer, wrapped_model_to_check):
 
         # with enforced DataParallel
         assert_flos_extraction(trainer, torch.nn.DataParallel(trainer.model))
+
+
+@require_torch
+@require_optuna
+class TrainerHyperParameterIntegrationTest(unittest.TestCase):
+    def setUp(self):
+        args = TrainingArguments(".")
+        self.n_epochs = args.num_train_epochs
+        self.batch_size = args.train_batch_size
+
+    def test_hyperparameter_search(self):
+        class MyTrialShortNamer(TrialShortNamer):
+            DEFAULTS = {"a": 0, "b": 0}
+
+        def hp_space(trial):
+            return {}
+
+        def model_init(trial):
+            if trial is not None:
+                a = trial.suggest_int("a", -4, 4)
+                b = trial.suggest_int("b", -4, 4)
+            else:
+                a = 0
+                b = 0
+            config = RegressionModelConfig(a=a, b=b, double_output=False)
+
+            return RegressionPreTrainedModel(config)
+
+        def hp_name(trial):
+            return MyTrialShortNamer.shortname(trial.params)
+
+        trainer_args = dict(model_init=model_init)
```

Comment:
If popped from the kwargs above, we can just pass model_init below, which is easier.

Reply:
OK,

Reply:
OK, but I am not really fan of poping stuff from dicts.
Here we have to know exactly which parameters are going to the trainer, and which are going to the trainer arguments.
If you forget one argument, or if two have the same name, it's becoming a mess.
For example, the RuntimeError I raised in default_compute_objective was because someone added "total_flos" to the metrics somewhere else but did not realize that default_compute_objective was popping values out of the metrics dictionary, and adding everything that was left in it. 
So the final eval metric in the end was the sum of the true eval metric and "total_flos", not really good.
So, yes I can use a dict.pop , but it makes some brittle code IMHO.

Reply:
I don't really see why this is similar to "total_flos" being added to the metrics (and the conclusion there is that it should never have been added to the metrics cause it's not one and it's polluting the logs when only the final value matters).

It's a helper function for some tests so there is little point debating on its API for hours. I was just expressing the fact I would have liked some consistency in this test file on how the defaults are changed but if it irks you that much, do what you want.

Reply:
Fixed in https://github.com/huggingface/transformers/pull/7935/commits/1c4d60817d922783f8bbac1764674918476dcbd9 .